edn/src/pretty_print.rs

```diff
@@ -37,70 +53,134 @@ impl Value {
+    where A: DocAllocator<'a> {
         match *self {
-            Value::Vector(ref vs) => self.bracket(pp, "[", vs, "]"),
-            Value::List(ref vs) => self.bracket(pp, "(", vs, ")"),
-            Value::Set(ref vs) => self.bracket(pp, "#{", vs, "}"),
-            Value::Map(ref vs) => {
-                let xs = vs.iter().rev().map(|(k, v)| k.as_doc(pp).append(pp.space()).append(v.as_doc(pp)).group()).intersperse(pp.space());
-                pp.text("{")
-                    .append(pp.concat(xs).nest(1))
-                    .append(pp.text("}"))
-                    .group()
-            }
-            Value::NamespacedSymbol(ref v) => pp.text(v.namespace.as_ref()).append("/").append(v.name.as_ref()),
-            Value::PlainSymbol(ref v) => pp.text(v.0.as_ref()),
-            Value::NamespacedKeyword(ref v) => pp.text(":").append(v.namespace.as_ref()).append("/").append(v.name.as_ref()),
-            Value::Keyword(ref v) => pp.text(":").append(v.0.as_ref()),
-            Value::Text(ref v) => pp.text("\"").append(v.as_ref()).append("\""),
-            _ => pp.text(self.to_string())
+            Value::Vector(ref vs) => self.bracket(alloc, style, "[", vs, "]"),
+            Value::List(ref vs) => self.bracket(alloc, style, "(", vs, ")"),
+            Value::Set(ref vs) => self.bracket(alloc, style, "#{", vs, "}"),
+            Value::Map(ref vs) => self.bracket_map(alloc, style, "{", vs, "}"),
+            Value::NamespacedSymbol(ref v) => alloc.text(v.namespace.as_ref()).append("/").append(v.name.as_ref()),
+            Value::PlainSymbol(ref v) => alloc.text(v.0.as_ref()),
+            Value::NamespacedKeyword(ref v) => alloc.text(":").append(v.namespace.as_ref()).append("/").append(v.name.as_ref()),
+            Value::Keyword(ref v) => alloc.text(":").append(v.0.as_ref()),
+            Value::Text(ref v) => alloc.text("\"").append(v.as_ref()).append("\""),
+            _ => alloc.text(self.to_string())
         }
     }
 }
 
 #[cfg(test)]
 mod test {
+    use super::*;
+
     use parse;
 
     #[test]
     fn test_pp_io() {
         let string = "$";
         let data = parse::value(string).unwrap().without_spans();
 
-        assert_eq!(data.write_pretty(40, &mut Vec::new()).is_ok(), true);
+        assert_eq!(data.write_pretty(Style::Clojure(MaxLineLength(40)), &mut Vec::new()).is_ok(), true);
     }
 
     #[test]
     fn test_pp_types_empty() {
         let string = "[ [ ] ( ) #{ } { }, \"\" ]";
         let data = parse::value(string).unwrap().without_spans();
 
-        assert_eq!(data.to_pretty(40).unwrap(), "[[] () #{} {} \"\"]");
+        assert_eq!(data.to_pretty(Style::Clojure(MaxLineLength(40))).unwrap(), "[[] () #{} {} \"\"]");
+    }
+
+    #[test]
+    fn test_pp_types_empty_expanded() {
+        let string = "[ [ ] ( ) #{ } { }, \"\" ]";
+        let data = parse::value(string).unwrap().without_spans();
+
+        // Ending whitespace with `\x20` to avoid editors from removing trailing
```

Comment:
I saw this in your earlier patch and didn't comment explicitly; I should have.  This is a really strong indicator that something is wrong.  We shouldn't be pretty printing output that we wouldn't commit to the tree!